tentacle/src/transports/tcp_base_listen.rs

```diff
@@ -0,0 +1,463 @@
+    self_mode: UpgradeMode,
+    #[cfg(feature = "tls")] config: TlsConfig,
+    global: Arc<crate::lock::Mutex<HashMap<SocketAddr, UpgradeMode>>>,
+    timeout: Duration,
+) -> Result<(Multiaddr, TcpBaseListenerEnum)> {
+    let addr = address.await?;
+    match multiaddr_to_socketaddr(&addr) {
+        Some(socket_address) => {
+            // Global register global listener upgrade mode
+            match global.clone().lock().entry(socket_address) {
+                Entry::Occupied(v) => {
+                    #[allow(unused_mut)]
+                    let mut tcp_base_addr: Multiaddr = socketaddr_to_multiaddr(socket_address);
+                    let listen_addr = match self_mode.to_enum() {
+                        UpgradeModeEnum::OnlyTcp => tcp_base_addr,
+                        #[cfg(feature = "ws")]
+                        UpgradeModeEnum::OnlyWs => {
+                            tcp_base_addr.push(Protocol::Ws);
+                            tcp_base_addr
+                        }
+                        #[cfg(feature = "tls")]
+                        UpgradeModeEnum::OnlyTLS => {
+                            match parse_tls_domain_name(&addr) {
+                                None => return Err(TransportErrorKind::NotSupported(addr)),
+                                Some(d) => {
+                                    tcp_base_addr.push(Protocol::Tls(Cow::Owned(d)));
+                                }
+                            }
+                            tcp_base_addr
+                        }
+                        #[allow(unreachable_patterns)]
+                        _ => unreachable!(),
+                    };
+                    v.get().combine(self_mode);
+                    return Ok((listen_addr, TcpBaseListenerEnum::Upgrade));
+                }
+                Entry::Vacant(v) => {
+                    v.insert(self_mode.clone());
+                }
+            }
+            let (local_addr, tcp) = tcp_listen(socket_address, tcp_config).await?;
+
+            #[allow(unused_mut)]
+            let mut tcp_base_addr: Multiaddr = socketaddr_to_multiaddr(local_addr);
+            let listen_addr = match self_mode.to_enum() {
+                UpgradeModeEnum::OnlyTcp => tcp_base_addr,
+                #[cfg(feature = "ws")]
+                UpgradeModeEnum::OnlyWs => {
+                    tcp_base_addr.push(Protocol::Ws);
+                    tcp_base_addr
+                }
+                #[cfg(feature = "tls")]
+                UpgradeModeEnum::OnlyTLS => {
+                    match parse_tls_domain_name(&addr) {
+                        None => return Err(TransportErrorKind::NotSupported(addr)),
+                        Some(d) => {
+                            tcp_base_addr.push(Protocol::Tls(Cow::Owned(d)));
+                        }
+                    }
+                    tcp_base_addr
+                }
+                #[allow(unreachable_patterns)]
+                _ => unreachable!(),
```

Comment:
Is is possible that `self_mode.to_enum()` is `UpgradeModeEnum::All`?

Reply:
Impossible, that each listen operation can only specify one protocol; to support multiple protocols, multiple listen operations are required.